File: program/src/state.rs
use {
	crate::error::WhitelistError,
	borsh::{BorshDeserialize, BorshSchema, BorshSerialize},
	solana_program::{
		entrypoint::ProgramResult,
		pubkey::Pubkey,
		sysvar::{clock::Clock, Sysvar},
	},
};

#[derive(BorshSerialize, BorshDeserialize, BorshSchema, Debug, PartialEq)]
pub struct Whitelist {
	pub bump: u8,
	pub authority: Pubkey,
	pub vault: Pubkey,
	pub treasury: Pubkey,
	pub mint: Pubkey,
	pub token_price: u64,
	pub buy_limit: u64,
	pub deposited: u64,
	pub whitelist_size: Option<u64>,
	pub allow_registration: bool,
	pub registration_start_timestamp: Option<i64>,
	pub registration_duration: Option<i64>,
	pub sale_start_timestamp: Option<i64>,
	pub sale_duration: Option<i64>,
}

impl Whitelist {
	pub const LEN: usize = 194;

	pub fn check_times(&self) -> ProgramResult {
		let clock = Clock::get()?;
		// Perform safety checks if a `registration_start_timestamp` is not `None`
		if let Some(registration_start_timestamp) = self.registration_start_timestamp {
			if registration_start_timestamp < clock.unix_timestamp {
				return Err(WhitelistError::InvalidRegistrationStartTime.into());
			}
		}

		// Perform safety checks if a `sale_start_timestamp` is not `None`
		if let Some(sale_start_timestamp) = self.sale_start_timestamp {
			if sale_start_timestamp < clock.unix_timestamp {
				return Err(WhitelistError::InvalidSaleStartTime.into());
			}
			if self
				.registration_start_timestamp
				.is_some_and(|t| t > sale_start_timestamp)
			{
				return Err(WhitelistError::SaleBeforeRegistration.into());
			}
		}

		Ok(())
	}

	pub fn check_sale_time(&self) -> ProgramResult {
		let clock = Clock::get()?;
		if self
			.sale_start_timestamp
			.is_some_and(|t| t >= clock.unix_timestamp)
		{
			Ok(())
		} else {
			Err(WhitelistError::SaleOngoing.into())
		}
	}
}

#[derive(BorshSerialize, BorshDeserialize, BorshSchema, Debug, PartialEq)]
pub struct Ticket {
	pub bump: u8,
    pub whitelist: Pubkey,
	pub owner: Pubkey,
	pub payer: Pubkey,
	pub allowance: u64,
	pub amount_bought: u64,
}

impl Ticket {
	pub const LEN: usize = 124;
}

File: js/src/state.ts
import { PublicKey } from "@solana/web3.js";
import { Schema, serialize } from "borsh";
import { Numberu64, Numberi64 } from "./numbers";


class WhitelistSchema {
    bump: number;
    authority: PublicKey;
    vault: PublicKey;
    treasury: PublicKey;
    mint: PublicKey;
    tokenPrice: Numberu64;
    buyLimit: Numberu64;
    deposited: Numberu64;
    whitelistSize: Numberu64;
    allowRegistration: boolean;
    registrationTimestamp: Numberi64;
    registrationDuration: Numberi64;
    saleTimestamp: Numberi64;
    saleDuration: Numberi64;

    constructor(fields: {
        bump: number,
        authority: PublicKey,
        vault: PublicKey,
        treasury: PublicKey,
        mint: PublicKey,
        tokenPrice: Numberu64,
        buyLimit: Numberu64,
        deposited: Numberu64
        whitelistSize: Numberu64,
        allowRegistration: boolean,
        registrationTimestamp: Numberi64
        registrationDuration: Numberi64,
        saleTimestamp: Numberi64,
        saleDuration: Numberi64,
    }) {
        this.bump = fields.bump;
        this.authority = fields.authority;
        this.vault = fields.vault;
        this.treasury = fields.treasury;
        this.mint = fields.mint;
        this.tokenPrice = fields.tokenPrice;
        this.buyLimit = fields.buyLimit;
        this.allowRegistration = fields.allowRegistration;
        this.registrationTimestamp = fields.registrationTimestamp;
        this.registrationDuration = fields.registrationDuration;
        this.saleTimestamp = fields.saleTimestamp;
        this.saleDuration = fields.saleDuration;
    }

    static schema: Schema = {
        struct: {
            bump: "u8",
            authority: { array: { type: "u8", len: 32 } },
            vault: { array: { type: "u8", len: 32 } },
            treasury: { array: { type: "u8", len: 32 } },
            mint: { array: { type: "u8", len: 32 } },
            tokenPrice: "u64",
            buyLimit: "u64",
            deposited: "u64",
            whitelistSize: "u64",
            allowRegistration: "bool",
            registrationTimestamp: "i64",
            registrationDuration: "i64",
            saleTimestamp: "i64",
            saleDuration: "i64",
        }
    };

    serialize(): Uint8Array {
        return serialize(WhitelistSchema.schema, this);
    }
}

class TicketSchema {
    bump: number;
    whitelist: PublicKey;
    owner: PublicKey;
    payer: PublicKey;
    allowance: PublicKey;
    amountBought: PublicKey;

    constructor(fields: {
        bump: number,
        whitelist: PublicKey,
        owner: PublicKey,
        payer: PublicKey,
        allowance: PublicKey,
        amountBought: PublicKey,
    }) {
        this.bump = fields.bump;
        this.whitelist = fields.whitelist;
        this.owner = fields.owner;
        this.payer = fields.payer;
        this.allowance = fields.allowance;
        this.amountBought = fields.amountBought;
    }

    static schema: Schema = {
        struct: {
            bump: "u8",
            whitelist: { array: { type: "u8", len: 32 } },
            owner: { array: { type: "u8", len: 32 } },
            payer: { array: { type: "u8", len: 32 } },
            allowance: "u64",
            amountBought: "u64",
        }
    };

    serialize(): Uint8Array {
        return serialize(TicketSchema.schema, this);
    }
}
